ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/TransactionsWriter.scala

```diff
@@ -103,23 +157,37 @@ private[dao] object TransactionsWriter extends TransactionsWriter {
       flatTransactionWitnessesBatch.foreach(_.execute())
       complementWitnessesBatch.foreach(_.execute())
 
+      val contractBatches = ContractsTable.prepareBatchInsert(
+        ledgerEffectiveTime = ledgerEffectiveTime,
+        transaction = transaction,
+        divulgedContracts = divulgedContracts,
+      )
+
+      for ((deleted, deleteContractsBatch) <- contractBatches.deletions) {
+        val deleteWitnessesBatch = contractWitnessesTable.prepareBatchDelete(deleted.toSeq)
+        require(deleteWitnessesBatch.nonEmpty, "Illegal: deleting witnesses without a contract")
```

Comment:
Shouldn't the error message say `deleting contracts without deleting witnesses`?

Reply:
No, in this line I'm checking that `deleteWitnessesBatch` is not empty, which it should not be, because we're deleting a contract and no contract should be added if if cannot be witnessed.

Reply:
Sorry, I don't follow. Maybe it is also a moot point because`deleteWitnessesBatch` will always be non-empty and 
maybe it's worth removing the optionality by making `PreparedBatches#deleted` and the parameter for `prepareBatchDelete` of type `Option[OneAnd[Set, ContractId]]`.

Reply:
What I'm saying is that this is an exceptional condition because I found a batch of witnesses that do not have any corresponding contracts. I understand that this constraint is currently implied by the underlying storage foreign key constraint, but I'd at least leave an `assert` here (the `require` is probably confusing, but the meaning was that of an `assert`).

I would prefer not pulling in `scalaz` just for one type, especially given the fact that it would imply some sort of `assert`-like behavior as soon as we pull data from the database.

If that's ok with you I would simply change this to an assert and possibly rephrase it to: `"Illegal: no witness found for contracts marked for deletion"`.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/5541/commits/3e9c6dec68fe7c51603a369e74fd3b5bfbe94f9e